parser/src/lexer.rs

```diff
@@ -516,803 +393,533 @@ where
             }
 
-            // Next, flush the indentation stack to zero.
-            while !self.indentations.is_empty() {
+            _ => {
+                self.at_begin_of_line = false;
+
+                return self.handle_indentation(indentation);
+            }
+        };
+
+        Ok(Some((token, self.token_range())))
+    }
+
+    fn handle_indentation(
+        &mut self,
+        indentation: Indentation,
+    ) -> Result<Option<Spanned>, LexicalError> {
+        let token = match self.indentations.current().try_compare(&indentation) {
+            // Dedent
+            Ok(Ordering::Greater) => {
                 self.indentations.pop();
-                self.emit((Tok::Dedent, TextRange::empty(tok_pos)));
+                self.pending_indentation = Some(indentation);
+
+                Some((Tok::Dedent, TextRange::empty(self.offset())))
             }
 
-            self.emit((Tok::EndOfFile, TextRange::empty(tok_pos)));
+            Ok(Ordering::Equal) => None,
+
+            // Indent
+            Ok(Ordering::Less) => {
+                self.indentations.push(indentation);
+                Some((Tok::Indent, self.token_range()))
+            }
+            Err(_) => {
+                return Err(LexicalError {
+                    error: LexicalErrorType::IndentationError,
+                    location: self.offset(),
+                });
+            }
+        };
+
+        Ok(token)
+    }
+
+    // Take a look at the next character, if any, and decide upon the next steps.
+    fn consume_normal(&mut self, first: char) -> Result<Option<Spanned>, LexicalError> {
+        if first.is_ascii() {
+            self.consume_ascii_character(first)
+        } else if is_unicode_identifier_start(first) {
+            let identifier = self.lex_identifier(first)?;
+            Ok(Some((identifier, self.token_range())))
+        } else if is_emoji_presentation(first) {
+            Ok(Some((
+                Tok::Name {
+                    name: first.to_string(),
+                },
+                self.token_range(),
+            )))
+        } else {
+            Err(LexicalError {
+                error: LexicalErrorType::UnrecognizedToken { tok: first },
+                location: self.token_start(),
+            })
+        }
+    }
+
+    fn consume_end(&mut self) -> Result<Spanned, LexicalError> {
+        // We reached end of file.
+        // First of all, we need all nestings to be finished.
+        if self.nesting > 0 {
+            return Err(LexicalError {
+                error: LexicalErrorType::Eof,
+                location: self.offset(),
+            });
         }
 
-        Ok(())
+        // Next, insert a trailing newline, if required.
+        if !self.at_begin_of_line {
+            self.at_begin_of_line = true;
+            Ok((Tok::Newline, TextRange::empty(self.offset())))
+        }
+        // Next, flush the indentation stack to zero.
+        else if self.indentations.pop().is_some() {
```

Comment:
is it intentional that this is not a while loop anymore?

Reply:
Yes, because the pending stack is gone.

The existing implementation writes any pending new line and dedents to the pending vector and then pops them one by one in the next method

The new implementation relies on the loop around next. It first returns the newline (and changes its internal state). The lexer is still at the end of the file for the next call and then returns the dedents one at a time until the stack is empty. It then finally returns the end of file token (forever)